active/0002-centalized-mgmt-resource-types-provider-versions.md

```diff
@@ -0,0 +1,247 @@
+
+**Example: Provider resolution error due as a result of `bicepconfig.json` merge semantics**
+
+Consider a Bicep project under folder `src` with the following files:
+
+- `src/main.bicep`
+
+```bicep
+provider foo as mainFooProvider
+```
+
+- `src/bicepconfig.json`
+
+```bicep
+{
+    "providers":{
+        "foo": {
+            "source": {
+                "registry": "mcr.microsoft.com/bicep/providers/az",
+                "version": "1.2.3",
+            },
+            "isDefaultImport": false
+        },
+        "bar" : {
+            "source": {
+                "registry": "mcr.microsoft.com",
+                "repository": "bicep/providers",
+                "name":
+                "version": "3.2.1"
+            },
+            "isDefaultImport": false
+        }
+    }
+}
+```
+
+- `src/modules/mod.bicep`
+
+```bicep
+provider bar as modBarProvider
+provider foo as modFooProvider
+```
+
+- `src/modules/bicepconfig.json`
+
+```bicep
+{
+     "experimentalFeaturesEnabled": {
+        "compileTimeImports": false
+     },
+}
+```
+
+The user may intuitively expect Bicep to resolve the provider details of `modFooProvider` by inspection of `src/bicepconfig.json` and for this project to compile successfully in a way that feels familiar with other programming languages where merge semantics are applied on code dependencies (e.g. NPMs `package.json`), however, Bicep config merge semantics uses the configuration file closest to the Bicep file in the directory hierarchy and merges them to the builtin config.
+
+In this case Bicep will try to resolve the `modFooProvider` details from `src/modules/bicepconfig.json`. The config doesn't specify a `providers` section, and the builtin config doesn't include a definition for provider `foo` and so the compiler cannot resolve the provider declaration statement and will emit a diagnostic.
+
+If we replace `foo` with `az` in the example above, Bicep resolves `modFooProvider`'s version to be the version of `az` specified in the builtin configuration and not the version `1.2.3` specified in `src/modules/mod.bicep` (inherited from `src/bicepconfig.json`) which is also not intuitive. The reson for this behavior is that current `bicepconfig.json` merge semantics merge `src/modules/bicepconfig.json` with the builtin config and the version applied would be the value specified in the builtin az provider version.
+
+This is especially problematic for 3rd party namespace identifiers specified in a `bicepconfig.json` up a file hierarchy, since they would result in a compilation error because they are unspecified in the builtin configuration.
+
+1. TODO: Potential drawback when trying to inspect module sources 
+<!--
```

Comment:
It would be good to get more clarity about the intention of the "inspect module sources" feature. If the goal is for the original source to be represented in full fidelity and compilable, then I agree this is important. If it's just to provide a quick reference (similar to firing up github or ADO code search), then this doesn't feel as important.

This is also technically orthogonal, as it's a problem regardless.

Reply:
TODO(asilverman): Move this discussion to the out-of-scope section